File: nnmcts/mcts/mcts.py

from functools import reduce

from nnmcts.mcts.nodes import Node


def mcts(node: Node, iters=100, show_execution_time=False):
  execution_times = []

  for _ in range(iters):
    perf = {}
    node.explore(perf)
    execution_times.append(perf)

  if show_execution_time:
    total_times = reduce(
      lambda x, y: {
        "traverse_time": x["traverse_time"] + y["traverse_time"],
        "rollout_time": x["rollout_time"] + y["rollout_time"],
        "update_time": x["update_time"] + y["update_time"],
      },
      execution_times,
      {"traverse_time": 0, "rollout_time": 0, "update_time": 0},
    )
    print(f"Traverse time: {total_times['traverse_time'] / iters:.4f}")
    print(f"Rollout time: {total_times['rollout_time'] / iters:.4f}")
    print(f"Update time: {total_times['update_time'] / iters:.4f}")

  policy = node.get_policy()
  next_node = node.get_most_visited()

  return next_node, policy


def collect_mcts_timing(node: Node, iters: int) -> dict[str, float]:
  totals = {"traverse_time": 0.0, "rollout_time": 0.0, "update_time": 0.0}

  for _ in range(iters):
    perf = {}
    node.explore(perf)
    for key in totals:
      totals[key] += perf.get(key, 0.0)

  count = max(iters, 1)
  return {key: value / count for key, value in totals.items()}

File: demo/src/lib/mcts/mcts.ts
import { NeuralNode, Node } from "./node";
import type { MctsEnvironment, MctsOptions, MctsResult } from "./types";

export function createRootNode<TMove, TEnv extends MctsEnvironment<TMove>>(
  environment: TEnv,
  useNeural = false,
): Node<TMove, TEnv> {
  const terminal = environment.isTerminal();
  return useNeural
    ? new NeuralNode<TMove, TEnv>(environment, terminal, null, null)
    : new Node<TMove, TEnv>(environment, terminal, null, null);
}

export async function mcts<TMove>(
  root: Node<TMove, MctsEnvironment<TMove>>,
  options: MctsOptions = {},
): Promise<MctsResult<TMove>> {
  const { iters = 100, showExecutionTime = false } = options;
  const executionTimes: Array<{
    traverseTime: number;
    rolloutTime: number;
    updateTime: number;
    createTime: number;
  }> = [];

  for (let i = 0; i < iters; i++) {
    const perf = {
      traverseTime: 0,
      rolloutTime: 0,
      updateTime: 0,
      createTime: 0,
    };
    await root.explore(perf);
    executionTimes.push(perf);
  }

  if (showExecutionTime) {
    const totals = executionTimes.reduce(
      (acc, perf) => ({
        traverseTime: acc.traverseTime + perf.traverseTime,
        rolloutTime: acc.rolloutTime + perf.rolloutTime,
        updateTime: acc.updateTime + perf.updateTime,
        createTime: acc.createTime + perf.createTime,
      }),
      {
        traverseTime: 0,
        rolloutTime: 0,
        updateTime: 0,
        createTime: 0,
      },
    );

    console.log(`Traverse time: ${(totals.traverseTime / iters).toFixed(4)}`);
    console.log(`Rollout time: ${(totals.rolloutTime / iters).toFixed(4)}`);
    console.log(`Update time: ${(totals.updateTime / iters).toFixed(4)}`);
    console.log(`Create time: ${(totals.createTime / iters).toFixed(4)}`);
  }

  const policy = root.getPolicy();
  const nextNode = root.getMostVisited();

  if (nextNode.action === null) {
    throw new Error("Most visited node has no action");
  }

  return {
    move: nextNode.action,
    policy,
  };
}

export { Node, NeuralNode } from "./node";
export type {
  ExplorePerf,
  MctsEnvironment,
  MctsOptions,
  MctsResult,
  UTTTEnvironment,
} from "./types";
